File: parser.py
from dataclasses import dataclass

from lexer import tokenize
from nodes import (
    Access,
    AncestorLookup,
    BinaryOp,
    Block,
    Call,
    CloneAttr,
    Conditional,
    Eager,
    Identifier,
    IntLit,
    Node,
    Override,
    Parent,
    SelfRef,
    StringLit,
)


@dataclass
class Definition:
    name: str
    expr: Node


class ParseError(Exception):
    pass


# Precedence for binary arithmetic operators (higher = tighter binding)
BinaryOpPrecedence = {
    "||": 3,
    "&&": 4,
    "==": 5,
    "!=": 5,
    "<=": 7,
    ">=": 7,
    "<": 7,
    ">": 7,
    "+": 10,
    "-": 10,
    "*": 20,
    "/": 20,
    "%": 20,
}


class Parser:
    def __init__(self, toks):
        self.toks = toks
        self.k = 0

    def peek(self, t=0):
        k = self.k + t
        return self.toks[k] if k < len(self.toks) else self.toks[-1]

    def match(self, *types):
        if self.peek().typ in types:
            tok = self.peek()
            self.k += 1
            return tok
        return None

    def expect(self, *types):
        tok = self.match(*types)
        if not tok:
            raise ParseError(
                f"Expected {types}, got {self.peek().typ} at {self.peek_position()}"
            )
        return tok

    def peek_position(self):
        return f"{self.peek().line}:{self.peek().col}"

    def consume_delims(self):
        while self.peek().typ in {","}:
            self.k += 1

    def parse_module(self) -> Block:
        defs = self.parse_defs_until({"EOF"})
        self.expect("EOF")
        return Block(defs)

    def parse_definition(self) -> Definition:
        name = self.expect("IDENT").val
        eq = self.expect("=", ":=", "<-")
        if eq.typ == "<-":
            attr = self.expect("IDENT").val
            return Definition(name, CloneAttr(attr=attr))
        expr = self.parse_expr()
        return Definition(name, expr if eq.typ == "=" else Eager(base=expr))

    def parse_defs_until(self, stop, allow_eager: bool = False) -> dict[str, Node]:
        defs = {}
        self.consume_delims()
        while self.peek().typ not in stop:
            start = self.peek_position()
            d = self.parse_definition()
            if d.name in defs:
                raise ParseError(f"Redefinition of {d.name!r} at {start}")
            elif not allow_eager and isinstance(d.expr, Eager):
                raise ParseError(
                    f"Eager definition is not allowed in this context at {start}"
                )
            defs[d.name] = d.expr
            self.consume_delims()
        return defs

    def parse_expr(self) -> Node:
        return self.parse_unary_or_ternary()

    def parse_unary_or_ternary(self) -> Node:
        """
        Parse ternary conditional: cond ? a : b
        `?:` has lower precedence than arithmetic and postfix.
        """
        node = self.parse_binary(0)
        if self.match("?"):
            true_expr = self.parse_expr()
            self.expect(":")
            false_expr = self.parse_expr()
            return Conditional(cond=node, a=true_expr, b=false_expr)
        return node

    def parse_binary(self, min_prec: int) -> Node:
        """
        Precedence-climbing parser for left-associative binary ops
        using BinaryOpPrecedence.
        """
        node = self.parse_postfix()
        while True:
            tok = self.peek()
            op = tok.typ
            if op not in BinaryOpPrecedence:
                break

            prec = BinaryOpPrecedence[op]
            if prec < min_prec:
                break

            # consume operator
            self.k += 1
            # left-associative: recurse with higher min_prec
            rhs = self.parse_binary(prec + 1)
            node = BinaryOp(x=node, op=op, y=rhs)
        return node

    def parse_postfix(self) -> Node:
        """
        Parse primary expression followed by:
        - .attr / .PARENT chains
        - UNWRAP
        - [ overrides ]
        - ( call(...) )
        """
        node = self.parse_primary()
        while True:
            if self.match("."):
                if self.match("PARENT"):
                    if isinstance(node, Parent):
                        node = Parent(node.depth + 1)
                    else:
                        raise ParseError("Unexpected ^ operator at {start}")
                    continue
                attr = self.expect("IDENT").val
                node = Access(node, attr)
            elif self.match("UNWRAP"):
                node = Access(node, "result")
            elif self.peek().typ == "[":
                self.expect("[")
                defs = self.parse_defs_until({"]"})
                self.expect("]")
                node = Override(node, defs)
            elif self.peek().typ == "(":
                self.expect("(")
                defs = self.parse_defs_until({")"}, allow_eager=True)
                self.expect(")")
                node = Call(node, defs)
            else:
                break
        return node

    def parse_primary(self) -> Node:
        start_pos = self.peek_position()
        token = self.expect(
            "{", "INT", "STRING", "SELF", "PARENT", "ANCESTOR", "IDENT", "("
        )
        if token.typ == "{":
            defs = self.parse_defs_until({"}"})
            self.expect("}")
            return Block(defs)
        elif token.typ == "INT":
            return IntLit(int(token.val))
        elif token.typ == "STRING":
            return StringLit(token.val)
        elif token.typ == "SELF":
            return SelfRef()
        elif token.typ == "PARENT":
            return Parent(1)
        elif token.typ == "ANCESTOR":
            self.expect(".")
            name = self.expect("IDENT").val
            return AncestorLookup(name)
        elif token.typ == "IDENT":
            return Identifier(token.val)
        elif token.typ == "(":
            # Parenthesized expression: (expr)
            expr = self.parse_expr()
            self.expect(")")
            return expr
        raise ParseError(f"Unexpected token {token.typ} at {start_pos}")


def parse_module(code: str):
    toks = tokenize(code)
    return Parser(toks).parse_module()

File: reflex/parser.go
package reflex

import (
	"fmt"
	"strconv"
)

type ParseError struct {
	Msg string
	Pos Pos
}

func (p *ParseError) Error() string {
	return fmt.Sprintf("%s at %s", p.Msg, p.Pos)
}

var binaryOpPrecedence = map[string]int{
	"||": 3,
	"&&": 4,
	"==": 5,
	"!=": 5,
	"<=": 7,
	">=": 7,
	"<":  7,
	">":  7,
	"+":  10,
	"-":  10,
	"*":  20,
	"/":  20,
	"%":  20,
}

type Parser struct {
	toks []*Token
	k    int
}

func NewParser(toks []*Token) *Parser {
	return &Parser{toks: toks}
}

func (p *Parser) peek() *Token {
	if p.k >= len(p.toks) {
		return p.toks[len(p.toks)-1]
	}
	return p.toks[p.k]
}

func (p *Parser) match(types ...string) *Token {
	t := p.peek()
	for _, x := range types {
		if t.Typ == x {
			p.k += 1
			return t
		}
	}
	return nil
}

func (p *Parser) expect(types ...string) (*Token, error) {
	if t := p.match(types...); t != nil {
		return t, nil
	}
	return nil, &ParseError{
		Msg: fmt.Sprintf("expected type in %#v but got %#v", types, p.peek().Typ),
		Pos: p.peek().Pos,
	}
}

func (p *Parser) consumeDelims() {
	for p.peek().Typ == "," {
		p.k += 1
	}
}

func (p *Parser) parseModule() (ASTNode, error) {
	startPos := p.peek().Pos
	defs, _, _, err := p.parseDefsUntil("EOF", false, false)
	if err != nil {
		return nil, err
	}
	if _, err := p.expect("EOF"); err != nil {
		return nil, err
	}
	return &ASTBlock{Pos: startPos, Defs: defs}, nil
}

func (p *Parser) parseDefsUntil(
	terminator string,
	allowAliases,
	allowEager bool,
) (defs map[string]ASTNode, aliases map[string]string, eager map[string]ASTNode, err error) {
	defs = map[string]ASTNode{}
	if allowAliases {
		aliases = map[string]string{}
	}
	if allowEager {
		eager = map[string]ASTNode{}
	}
	for p.peek().Typ != terminator {
		name, err := p.expect("IDENT")
		if err != nil {
			return nil, nil, nil, err
		}
		t := p.peek()
		if t.Typ == "=" {
			val, err := p.parseExpr()
			if err != nil {
				return nil, nil, nil, err
			}
			defs[name.Val] = val
		} else if allowAliases && t.Typ == "<-" {
			ident, err := p.expect("IDENT")
			if err != nil {
				return nil, nil, nil, err
			}
			aliases[name.Val] = ident.Val
		} else if allowEager && t.Typ == ":=" {
			val, err := p.parseExpr()
			if err != nil {
				return nil, nil, nil, err
			}
			eager[name.Val] = val
		} else {
			return nil, nil, nil, &ParseError{Msg: fmt.Sprintf("unexpected token %#v inside definition", t.Typ), Pos: t.Pos}
		}
		p.consumeDelims()
	}
	return
}

func (p *Parser) parseExpr() (ASTNode, error) {
	node, err := p.parseBinary(0)
	if err != nil {
		return nil, err
	}
	if t := p.match("?"); t != nil {
		trueExpr, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		if _, err := p.expect(":"); err != nil {
			return nil, err
		}
		falseExpr, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		return &ASTTernary{Pos: t.Pos, Cond: node, IfTrue: trueExpr, IfFalse: falseExpr}, nil
	}
	return node, nil
}

func (p *Parser) parseBinary(minPrec int) (ASTNode, error) {
	node, err := p.parsePostfix()
	if err != nil {
		return nil, err
	}
	for {
		t := p.peek()
		op := t.Typ
		prec, ok := binaryOpPrecedence[op]
		if !ok || prec < minPrec {
			break
		}
		p.k += 1
		rhs, err := p.parseBinary(minPrec + 1)
		if err != nil {
			return nil, err
		}
		opName := map[string]string{
			"==": "eq",
			"!=": "ne",
			"<":  "lt",
			">":  "gt",
			">=": "ge",
			"<=": "le",
			"+":  "add",
			"-":  "sub",
			"/":  "div",
			"*":  "mul",
			"%":  "mod",
			"&&": "logical_and",
			"||": "logical_or",
		}[op]
		node = &ASTBinaryOp{
			Pos:    t.Pos,
			OpName: opName,
			X:      node,
			Y:      rhs,
		}
	}
	return node, nil
}

func (p *Parser) parsePostfix() (ASTNode, error) {
	node, err := p.parsePrimary()
	if err != nil {
		return nil, err
	}
	for {
		if p.match(".") != nil {
			startPos := p.peek().Pos
			if p.match("PARENT") != nil {
				if oldParent, ok := node.(*ASTParent); ok {
					node = &ASTParent{Pos: oldParent.Pos, Depth: oldParent.Depth + 1}
				} else {
					return nil, &ParseError{Msg: "invalid parent chaining", Pos: startPos}
				}
			} else {
				attr, err := p.expect("IDENT")
				if err != nil {
					return nil, err
				}
				node = &ASTAccess{Pos: attr.Pos, Base: node, Attr: attr.Val}
			}
		} else if uw := p.match("UNWRAP"); uw != nil {
			node = &ASTAccess{Pos: uw.Pos, Base: node, Attr: "result"}
		} else if open := p.match("["); open != nil {
			defs, aliases, _, err := p.parseDefsUntil("]", true, false)
			if err != nil {
				return nil, err
			}
			if _, err := p.expect("]"); err != nil {
				return nil, err
			}
			node = &ASTOverride{Pos: open.Pos, Base: node, Defs: defs, Aliases: aliases}
		} else if open := p.match("("); open != nil {
			defs, _, eager, err := p.parseDefsUntil("]", false, true)
			if err != nil {
				return nil, err
			}
			if _, err := p.expect("("); err != nil {
				return nil, err
			}
			node = &ASTCall{Pos: open.Pos, Base: node, Defs: defs, Eager: eager}
		} else {
			break
		}
	}
	return node, nil
}

func (p *Parser) parsePrimary() (ASTNode, error) {
	startPos := p.peek().Pos
	tok, err := p.expect("{", "INT", "STRING", "SELF", "PARENT", "ANCESTOR", "IDENT", "(")
	if err != nil {
		return nil, err
	}
	switch tok.Typ {
	case "{":
		defs, _, _, err := p.parseDefsUntil("}", false, false)
		if err != nil {
			return nil, err
		}
		return &ASTBlock{Pos: startPos, Defs: defs}, nil
	case "INT":
		value, err := strconv.ParseInt(tok.Val, 10, 64)
		if err != nil {
			return nil, &ParseError{Msg: err.Error(), Pos: startPos}
		}
		return &ASTIntLit{Pos: startPos, Value: value}, nil
	case "STRING":
		return &ASTStrLit{Pos: startPos, Value: tok.Val}, nil
	case "SELF":
		return &ASTSelfRef{Pos: startPos}, nil
	case "PARENT":
		return &ASTParent{Pos: startPos, Depth: 1}, nil
	case "ANCESTOR":
		if _, err := p.expect("."); err != nil {
			return nil, err
		}
		name, err := p.expect("IDENT")
		if err != nil {
			return nil, err
		}
		return &ASTAncestorLookup{Pos: startPos, Attr: name.Val}, nil
	case "IDENT":
		return &ASTIdentifier{Pos: startPos, Name: tok.Val}, nil
	case "(":
		expr, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		if _, err := p.expect(")"); err != nil {
			return nil, err
		}
		return expr, nil
	default:
		panic("unreachable")
	}
}
